File: src/main/java/uk/co/reecedunn/intellij/plugin/xquery/resolve/reference/XQueryFunctionNameReference.java
package uk.co.reecedunn.intellij.plugin.xquery.resolve.reference;

import com.intellij.openapi.util.TextRange;
import com.intellij.psi.PsiElement;
import com.intellij.psi.PsiReferenceBase;
import org.jetbrains.annotations.NotNull;
import uk.co.reecedunn.intellij.plugin.xquery.ast.xquery.*;
import uk.co.reecedunn.intellij.plugin.core.functional.Option;
import uk.co.reecedunn.intellij.plugin.xquery.psi.XQueryPrologResolver;

import static uk.co.reecedunn.intellij.plugin.core.functional.PsiTreeWalker.children;

public class XQueryFunctionNameReference extends PsiReferenceBase<XQueryEQName> {
    public XQueryFunctionNameReference(XQueryEQName element, TextRange range) {
        super(element, range);
    }

    @Override
    public PsiElement resolve() {
        return getElement().resolvePrefixNamespace().map((ns) -> {
            if (ns.getDeclaration() instanceof XQueryPrologResolver) {
                XQueryPrologResolver provider = (XQueryPrologResolver) ns.getDeclaration();
                XQueryProlog prolog = provider.resolveProlog();
                if (prolog == null) {
                    return null;
                }

                return Option.of(getElement().getLocalName()).map(PsiElement::getText).map((localName) -> {
                    PsiElement parent = getElement().getParent();
                    int arity = 0;
                    if (parent instanceof XQueryFunctionCall) {
                        arity = ((XQueryFunctionCall)parent).getArity();
                    } else if (parent instanceof XQueryNamedFunctionRef) {
                        arity = ((XQueryNamedFunctionRef)parent).getArity();
                    } else if (parent instanceof XQueryArrowFunctionSpecifier) {
                        arity = ((XQueryArrowFunctionSpecifier)parent).getArity();
                    }

                    PsiElement annotation = prolog.getFirstChild();
                    while (annotation != null) {
                        if (annotation instanceof XQueryAnnotatedDecl) {
                            XQueryFunctionDecl functionDecl = children(annotation).findFirst(XQueryFunctionDecl.class).getOrElse(null);
                            if (functionDecl != null) {
                                XQueryEQName functionName = children(functionDecl).findFirst(XQueryEQName.class).getOrElse(null);
                                Option<PsiElement> functionLocalName = functionName == null ? Option.none() : Option.of(functionName.getLocalName());
                                if (functionLocalName.map(PsiElement::getText).map((name) -> name.equals(localName)).getOrElse(false)) {
                                    if (functionDecl.getArity() == arity) {
                                        return functionName;
                                    }
                                }
                            }
                        }

                        annotation = annotation.getNextSibling();
                    }
                    return null;
                }).getOrElse(null);
            }
            return null;
        }).getOrElse(null);
    }

    @NotNull
    @Override
    public Object[] getVariants() {
        return new Object[0];
    }
}

File: src/main/java/uk/co/reecedunn/intellij/plugin/xquery/resolve/reference/XQueryFunctionNameReference.kt
/*
 * Copyright (C) 2016-2017 Reece H. Dunn
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package uk.co.reecedunn.intellij.plugin.xquery.resolve.reference

import com.intellij.openapi.util.TextRange
import com.intellij.psi.PsiElement
import com.intellij.psi.PsiReferenceBase
import uk.co.reecedunn.intellij.plugin.core.extensions.children
import uk.co.reecedunn.intellij.plugin.xquery.ast.xquery.*

class XQueryFunctionNameReference(element: XQueryEQName, range: TextRange) : PsiReferenceBase<XQueryEQName>(element, range) {
    override fun resolve(): PsiElement? {
        val prologs: Sequence<XQueryProlog> = element.resolvePrefixNamespace().filterIsInstance<XQueryProlog>()

        val localName = element.localName?.text
        val functions: Sequence<XQueryFunctionDecl> = prologs.flatMap { prolog ->
            prolog.children().filterIsInstance<XQueryAnnotatedDecl>().map { annotation ->
                val function = annotation.children().filterIsInstance<XQueryFunctionDecl>().firstOrNull()
                if (function?.text?.equals(localName) == true) {
                    function
                } else {
                    null
                }
            }
        }.filterNotNull()

        val parent = element.parent
        val arity = when (parent) {
            is XQueryFunctionCall -> parent.arity
            is XQueryNamedFunctionRef -> parent.arity
            is XQueryArrowFunctionSpecifier -> parent.arity
            else -> -1
        }

        return functions.firstOrNull { f -> f.arity == arity }
    }

    override fun getVariants(): Array<Any> {
        return arrayOf()
    }
}
